WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt

```diff
@@ -203,6 +209,7 @@ class NewSitePreviewViewModel @Inject constructor(
              */
             withContext(MAIN) {
                 if (uiState.value !is SitePreviewContentUiState) {
+                    tracker.trackPreviewWebviewShown()
```

Comment:
I am a little confused by this. Isn't the one in `onUrlLoaded` the correct place to call this tracker? We simply show the shimmer animation here and not the actual web view. When I debug it, the event is triggered twice. Maybe we should have a different tracker here or simply remove this one. What do you think?

Reply:
There are two events `trackPreviewWebviewShown` and `trackPreviewWebviewFullyLoaded`.
1. `trackPreviewWebviewShown` -> indicates the webview has been dispalyed. Even when it's empty/shimmer animation.
2. `trackPreviewWebviewFullyLoaded` -> indicates the newly created site is displayed in the webview.


There are two scenarios
a) We show a progress animation for 1s -> we display a webview with the shimmer animation-> the url load completes.
b) The url load completes within the first second -> there is no shimmer animation.
As far as I understand the events descriptions, we want to log `trackPreviewWebviewShown` in both scenarios.


However, the event should never be invoked twice. I'm looking at the code and it shouldn't be happening 😕

Reply:
I think it's a case of timing, I am pretty sure it has happened to me during my debug session. I _think_ we just need to move this https://github.com/wordpress-mobile/WordPress-Android/blob/issue/8509-site-creation-analytics/WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt#L228-L230 inside this `if` https://github.com/wordpress-mobile/WordPress-Android/blob/issue/8509-site-creation-analytics/WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt#L235. That'll make it so that only `trackPreviewWebviewShown` if the webview is shown for the first time.

Checking out #8870 which addresses [this issue](https://github.com/wordpress-mobile/WordPress-Android/issues/8824#issuecomment-448965637) might be helpful to understand why this is happening. In short, the webview might be shown either because a 1 second timeout is finished, or because it's fully loaded, whichever comes first. I guess we could add a helper method for this and add the tracking code there to make it easier to understand.

Let me know if that makes sense or if I can clarify it further. (or if I am missing something 🙈 )

Reply:
> Checking out #8870 which addresses this issue might be helpful to understand why this is happening. In short, the webview might be shown either because a 1 second timeout is finished, or because it's fully loaded, whichever comes first. I guess we could add a helper method for this and add the tracking code there to make it easier to understand.

Isn't this exactly what I tried to explain [here](https://github.com/wordpress-mobile/WordPress-Android/pull/8998/files#r248965155)? :D One of us is obviously missing something. It's probably me, however, I'd like to have a full understanding why it's happening before I make the change.

>I think it's a case of timing, I am pretty sure it has happened to me during my debug session. I think we just need to move this https://github.com/wordpress-mobile/WordPress-Android/blob/issue/8509-site-creation-analytics/WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt#L228-L230 inside this if https://github.com/wordpress-mobile/WordPress-Android/blob/issue/8509-site-creation-analytics/WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt#L235. That'll make it so that only trackPreviewWebviewShown if the webview is shown for the first time.

If I move the whole `if (uiState.value is SitePreviewFullscreenProgressUiState)` inside `if (uiState.value !is SitePreviewContentUiState)` I think nothing will change at all. As the first if already guarantees we are showing the fullscreen progress screen and not the shimmer animation.

If I move just `tracker.trackPreviewWebviewShown()` inside the `if (uiState.value !is SitePreviewContentUiState)` it'll track the event even when I'm going from `SitePreviewLoadingShimmerState -> SitePreviewContentUiState`. Which afaiu would log the event for the second time as it has already been logged when we showed the ShimmerAnimation.
 
However, right now it should log the event only when it's going from `SitePreviewFullscreenProgressUiState -> SitePreviewContentUiState` which means, the shimmer animations hasn't been shown.

Reply:
> If I move the whole if (uiState.value is SitePreviewFullscreenProgressUiState) inside if (uiState.value !is SitePreviewContentUiState) I think nothing will change at all. As the first if already guarantees we are showing the fullscreen progress screen and not the shimmer animation.

🤦‍♂️ 

@malinajirka I am really sorry for wasting your time. I have been looking at the code and I can't figure out how I could get a double event at all. You're absolutely right that above ^ suggestion was just stupid. I have also tested the flow 3 times and didn't get the double event at all.

I think I may have confused two different events when I first tested everything and thought it was a double event. Again, really sorry about this back and forth!